torch/csrc/autograd/engine.cpp

```diff
@@ -277,10 +306,23 @@ auto Engine::thread_main(
   // Why the test on graph_task->outstanding_tasks_?  See
   // Note [Reentrant backwards]
   while (!reentrant_thread || graph_task->outstanding_tasks_ > 0) {
-    NodeTask task = queue->pop();
+    std::unique_ptr<NodeTask> task{nullptr};
+    // If graph_task is not empty, the current thread is reentrant thread
+    if (graph_task) {
+      // For the case where there're more re-entrant threads than outstanding
+      // tasks, then some thread that couldn't grab the task will hang on
+      // trying to pop from readyQueue and hold the mutex_, give a timeout to
+      // pop (return null if timeout) to ensure they don't hang forever
```

Comment:
This still seems wrong to me, on multiple levels.

1. Traditionally, if you hold a lock and then want to do a blocking operation, you're supposed to release the lock before you block (and make sure you have restored all locking invariants when you lock). So this explanation doesn't make sense to me: it seems like you've observed a deadlock... and then added a timeout(!!) of all things to fix the problem. Don't do that!
2.  What if we have run out of tasks in this queue, but we are still making progress on other threads, it seems like you're going to end up with all of the other threads spin-waiting on the empty queue (because they aren't blocking). The degree to which their spinning is made slightly better by however big you make the timeout... but, echoing (1) again, the right way to do the synchronization here is to NOT fix your problem with timeouts.
3. @albanD mentioned to me that you don't have a good way to wake up all the threads when it's time to shut down, which is why you kick them out of blocking every once in a while so they can check for the other condition. This feels self inflicted to me: `condition_variable` explicitly has `notify_one` and `notify_all` for this exact case. You should be able to wake up all the threads when it's time to shut down. If the current API doesn't let you do that, fix the API so you can!